server/sonar-ce/src/main/java/org/sonar/ce/notification/ReportAnalysisFailureNotificationExecutionListener.java

```diff
@@ -0,0 +1,89 @@
+import org.sonar.ce.queue.CeTask;
+import org.sonar.ce.queue.CeTaskResult;
+import org.sonar.ce.taskprocessor.CeWorker;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.RowNotFoundException;
+import org.sonar.db.ce.CeActivityDto;
+import org.sonar.db.ce.CeTaskTypes;
+import org.sonar.db.component.ComponentDto;
+import org.sonar.server.notification.NotificationService;
+
+import static java.lang.String.format;
+import static java.util.Collections.singleton;
+
+public class ReportAnalysisFailureNotificationExecutionListener implements CeWorker.ExecutionListener {
+  private final NotificationService service;
+  private final DbClient dbClient;
+  private final ReportAnalysisFailureNotificationSerializer taskFailureNotificationSerializer;
+  private final System2 system2;
+
+  public ReportAnalysisFailureNotificationExecutionListener(NotificationService service, DbClient dbClient,
+    ReportAnalysisFailureNotificationSerializer taskFailureNotificationSerializer, System2 system2) {
+    this.service = service;
+    this.dbClient = dbClient;
+    this.taskFailureNotificationSerializer = taskFailureNotificationSerializer;
+    this.system2 = system2;
+  }
+
+  @Override
+  public void onStart(CeTask ceTask) {
+    // nothing to do
+  }
+
+  @Override
+  public void onEnd(CeTask ceTask, CeActivityDto.Status status, @Nullable CeTaskResult taskResult, @Nullable Throwable error) {
+    if (status == CeActivityDto.Status.SUCCESS || !ceTask.getType().equals(CeTaskTypes.REPORT)) {
+      return;
+    }
+
+    String projectUuid = ceTask.getComponentUuid();
+    if (projectUuid != null && service.hasProjectSubscribersForTypes(projectUuid, singleton(ReportAnalysisFailureNotification.TYPE))) {
+      try (DbSession dbSession = dbClient.openSession(false)) {
+        ComponentDto projectDto = dbClient.componentDao().selectOrFailByUuid(dbSession, projectUuid);
+        CeActivityDto ceActivityDto = dbClient.ceActivityDao().selectByUuid(dbSession, ceTask.getUuid())
+          .orElseThrow(() -> new RowNotFoundException(format("CeActivityDao with uuid %s not found", ceTask.getUuid())));
+        ReportAnalysisFailureNotification taskFailureNotification = buildNotification(ceActivityDto, projectDto, error);
+        service.deliver(taskFailureNotificationSerializer.toNotification(taskFailureNotification));
+      }
+    }
+  }
+
+  private ReportAnalysisFailureNotification buildNotification(CeActivityDto ceActivityDto, ComponentDto projectDto, @Nullable Throwable error) {
+    Long executedAt = ceActivityDto.getExecutedAt();
+    return new ReportAnalysisFailureNotification(
+      new ReportAnalysisFailureNotification.Project(
+        projectDto.uuid(),
+        projectDto.getKey(),
+        projectDto.name(),
+        projectDto.getBranch()),
+      new ReportAnalysisFailureNotification.Task(
+        ceActivityDto.getUuid(),
+        ceActivityDto.getSubmittedAt(),
+        executedAt == null ? system2.now() : executedAt),
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this use of "now"; it is deprecated. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)